x-pack/plugins/apm/server/lib/services/annotations/index.ts

```diff
@@ -3,112 +3,66 @@
   setup,
   serviceName,
-  environment
+  environment,
+  annotationsClient
 }: {
   serviceName: string;
   environment?: string;
   setup: Setup & SetupTimeRange;
+  annotationsClient?: ScopedAnnotationsClient;
 }) {
-  const { start, end, client, indices } = setup;
-
-  const filter: ESFilter[] = [
-    { term: { [PROCESSOR_EVENT]: 'transaction' } },
-    { range: rangeFilter(start, end) },
-    { term: { [SERVICE_NAME]: serviceName } }
-  ];
+  async function getStoredAnnotations(): Promise<Annotation[]> {
+    if (!annotationsClient) {
+      return [];
+    }
 
-  if (environment) {
-    filter.push({ term: { [SERVICE_ENVIRONMENT]: environment } });
-  }
-
-  const versions =
-    (
-      await client.search({
-        index: indices['apm_oss.transactionIndices'],
-        body: {
-          size: 0,
-          track_total_hits: false,
-          query: {
-            bool: {
-              filter
-            }
-          },
-          aggs: {
-            versions: {
-              terms: {
-                field: SERVICE_VERSION
-              }
-            }
-          }
+    const response = await annotationsClient.search({
+      start: setup.start,
+      end: setup.end,
+      annotation: {
+        type: 'deployment'
+      },
+      tags: ['apm'],
+      filter: {
+        term: {
+          [SERVICE_NAME]: serviceName
         }
-      })
-    ).aggregations?.versions.buckets.map(bucket => bucket.key) ?? [];
-
-  if (versions.length > 1) {
-    const annotations = await Promise.all(
-      versions.map(async version => {
-        const response = await client.search({
-          index: indices['apm_oss.transactionIndices'],
-          body: {
-            size: 0,
-            query: {
-              bool: {
-                filter: filter
-                  .filter(esFilter => !Object.keys(esFilter).includes('range'))
-                  .concat({
-                    term: {
-                      [SERVICE_VERSION]: version
-                    }
-                  })
-              }
-            },
-            aggs: {
-              first_seen: {
-                min: {
-                  field: '@timestamp'
-                }
-              }
-            },
-            track_total_hits: false
-          }
-        });
+      }
+    });
 
-        const firstSeen = response.aggregations?.first_seen.value;
+    return response.hits.hits.map(hit => {
+      return {
+        type: AnnotationType.VERSION,
+        id: hit._id,
+        time: new Date(hit._source['@timestamp']).getTime(),
+        text: hit._source.message
+      };
+    });
+  }
 
-        if (!isNumber(firstSeen)) {
-          throw new Error(
-            'First seen for version was unexpectedly undefined or null.'
-          );
-        }
+  const [derivedAnnotations, storedAnnotations] = await Promise.all([
+    getDerivedServiceAnnotations({
+      setup,
+      serviceName,
+      environment
+    }),
+    getStoredAnnotations()
+  ]);
```

Comment:
I might change this to something that will start both requests, but return as soon as the stored annotations are resolved. The aggregation to get the derived annotations can be quite costly.

Reply:
done